blog/2020/05/rr.md

```diff
@@ -0,0 +1,452 @@
+[technical report](https://arxiv.org/abs/1610.02144) for a deeper overview of
+some of `rr`'s design points (though even that only scratches the surface).
+
+One final thought for this section is the origin of the term "time-traveling"
+debugger. It stems from the mode of analysis that a system like this enables.
+In a traditional debugger, it is possible to execute one instruction at a time,
+and step from memory state to memory state in the forwards direction. Systems
+like `rr` allow the opposite during replay: Step backwards through the state of
+the system. Under the hood this is accomplished by playing forward from the
+beginning (or more likely some intermediate checkpoint created for performance
+reasons), until the previous state is reached. However, to the end user, the
+illusion of going backwards in time is presented and an extremely useful mental
+model for debugging.
+
+# Performance considerations
+
+While the previous section described the operating principle of `rr`, it does
+not do justice to the reason why `rr` works so well. That reason is simple:
+performance. Overhead for recording of single threaded processes
+is generally below 2x, most often between 2% and 50% (lower for purely
+numerical calculations, higher for workloads that interact with the OS).
+Recording multi-threaded processes is tougher. By default `rr` serializes
+execution of tasks that access the same shared memory. In generality this is
+required for correctness, since it is not possible to observe and record the
+interleaving of memory operations to shared memory spaces. However, there is
+hope. If the multi-threaded program is well-behaved (or mostly well-behaved)
+in that communication happens by message passing rather than shared memory, or
+if accesses to shared memory are explicitly synchronized in a recordable fashion,
+multi-threaded recording becomes possible. `rr` currently supports the former,
+if the messages are passed via operating system pipes. We are interested in
+techniques for the latter and there is some interesting academic work in this
+direction, but nothing close to production ready. A complete solution would
+probably combine compiler-based analysis of safe regions with dynamic
+enforcement of safety for regions where this cannot be proven by the compiler.
+However, that is for future work. For the moment, recording multi-threaded,
+shared-memory applications will likely incur overhead linear in the number of
+threads. It is thus a good idea to try and reproduce any issues at low core count.
+
+With that caveat in mind. Here is some real benchmark numbers of overhead for
+the julia test suite. The tests suite mostly uses message-passing based
+parallelism for running multiple tests in parallel. Only the `threads` test
+incurs the shared memory overhead penalty. Here are the measurements:
+
+[data]
+
+The total compressed size of a trace for a full run of the test suite (about
```

Comment:
```suggestion
Surprisingly too, your traces may often be quite small! The total compressed size of a trace for a full run of the test suite (about
```

Reply:
Fair, people may not have accurate calibrated expectations ;). Applied.